rustls/src/tls12/mod.rs

```diff
@@ -198,15 +199,16 @@ impl fmt::Debug for Tls12CipherSuite {
 }
 
 /// TLS1.2 per-connection keying material
-pub(crate) struct ConnectionSecrets {
+pub(crate) struct ConnectionSecrets<C: CryptoProvider> {
     pub(crate) randoms: ConnectionRandoms,
     suite: &'static Tls12CipherSuite,
     pub(crate) master_secret: [u8; 48],
+    backend: PhantomData<C>,
 }
 
-impl ConnectionSecrets {
+impl<C: CryptoProvider> ConnectionSecrets<C> {
```

Comment:
Here I've made this `impl` block generic over `CryptoProvider` **but** several of the impl functions are in-fact reliant on `Ring` in particular. E.g. `make_cipher_pair` takes a `aead::Algorithm` parameter.

I initially had two `impl` blocks, one that was `<C: CryptoProvider>` and one that was `<Ring>`, but then I had compiler errors from call-sites that were `<C: CryptoProvider>` constrained but needed to call these functions.

My thinking was we'd want to include these functions in the `<C: CryptoProvider>` block until we have more traits that we can use to pull replace the Ring specific bits. This may/may not match what you had in mind!